dart/lib/src/http_client/failed_request_client.dart

```diff
@@ -0,0 +1,245 @@
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+class FailedRequestClient extends BaseClient {
+  FailedRequestClient({
+    this.maxRequestBodySize = MaxRequestBodySize.small,
+    this.failedRequestStatusCodes = const [],
+    this.captureFailedRequests = true,
+    Client? client,
+    Hub? hub,
+  })  : _hub = hub ?? HubAdapter(),
+        _client = client ?? Client();
+
+  final Client _client;
+  final Hub _hub;
+
+  /// Configures wether to record exceptions for failed requests.
+  /// Examples for captures exceptions are:
+  /// - In an browser environment this can be requests which fail because of CORS.
+  /// - In an mobile or desktop application this can be requests which failed
+  ///   because the connection was interrupted.
+  final bool captureFailedRequests;
+
+  /// Configures up to which size request bodies should be included in events.
+  /// This does not change wether an event is captured.
+  final MaxRequestBodySize maxRequestBodySize;
+
+  /// Describes which HTTP status codes should be considered as a failed
+  /// requests.
+  ///
+  /// Per default no status code is considered a failed request.
+  final List<SentryStatusCode> failedRequestStatusCodes;
+
+  @override
+  Future<StreamedResponse> send(BaseRequest request) async {
+    int? statusCode;
+    Object? exception;
+    StackTrace? stackTrace;
+
+    final stopwatch = Stopwatch();
+    stopwatch.start();
+
+    try {
+      final response = await _client.send(request);
+      statusCode = response.statusCode;
+      return response;
+    } catch (e, st) {
+      exception = e;
+      stackTrace = st;
+      rethrow;
+    } finally {
+      stopwatch.stop();
+
+      // If captureFailedRequests is true, there statusCode is null.
+      // So just one of these blocks can be called.
+
+      if (captureFailedRequests && exception != null) {
+        await _captureEvent(
+          exception: exception,
+          stackTrace: stackTrace,
+          request: request,
+          requestDuration: stopwatch.elapsed,
+          withPii: true,
+        );
+      }
+
+      if (failedRequestStatusCodes.containsStatusCode(statusCode)) {
+        // Capture an exception if the status code is considered bad
+        await _captureEvent(
+          request: request,
+          reason: failedRequestStatusCodes.toDescription(),
+          requestDuration: stopwatch.elapsed,
+          withPii: true,
+        );
+      }
+    }
+  }
+
+  @override
+  void close() {
+    // See https://github.com/getsentry/sentry-dart/pull/226#discussion_r536984785
+    _client.close();
+  }
+
+  // See https://develop.sentry.dev/sdk/event-payloads/request/
+  Future<void> _captureEvent({
+    Object? exception,
+    StackTrace? stackTrace,
+    String? reason,
+    required Duration requestDuration,
+    required BaseRequest request,
+    required bool withPii,
```

Comment:
this should be read from `options.sendDefaultPii`, the thing is, there's no way to access it, in this case, you might need to expose the `Hub#getOptions` like Java does https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/Hub.java#L438-L441

so instead of taking  `withPii` as param, you actually call `hub.getOptions.sendDefaultPii`

Reply:
Yeah, I thought so too.
How would that work for the HubAdapter?
The HubAdapter only passes through to the static methods.

Reply:
damn, Java does `Sentry.getCurrentHub().getOptions()` and we just removed `getCurrentHub` cus its not part of the unified API but it'd be fine as a private method, wondering if it makes sense to rollback https://github.com/getsentry/sentry-dart/pull/490/files and mark it as internal, and use it to access the options.

I don't see another way of sharing the `defaultDefaultPii` flag, maybe @bruno-garcia has ideas?

Reply:
We could also just have a `includePii` for the constructor. I mean the user has to use the client manually anyway.

Though I'm not saying it's necessarily better 🤷‍♂️

Reply:
works for me